Use rendered imgui in code for settings (and more...)

mgavin opened the pull request:
Added IMGUI libraries to the plugin, inherited `pluginsettings`, rewrote settings into plugin
  - Removes need for `SettingsFile.cpp` and `writeSettingsFile` function
  - Enables for more customization of the plugin menu layout and look
  - Added a more reactive way to capture key input
    - RL's HWND is captured and used to clear mouse inputs that get stuck due to bakkesmod. (it still happens but it's an extra protection 🙄)
    - opened menus are closed and reopened after a bind has been captured
    - rewind and fast forward are the only ones able to capture inputs that have axes
    - removes the need to loop through all keys to find which one was captured
  - made links out of the links in the text at the bottom
 
Decoupled rewind and fast-foward from the `Steer` parameter of `ControllerInput`
  - added a value to the `rewindState` that tracks it
  - code hooked during rewind mode allows for an axis input on one and another input on another

mgavin committed:
Updated the platform toolset from `v142` to `v143`.
Enabled address sanitation, removed console entry point, disabled bakkesmod-caused warning `4251`, enabled string pooling
Updated the set language standard to `stdcpplatest`
Removed the reliance (and files) on the `fmt` library in favor of `std::format`
Added IMGUI files to the project's filters for building alongside the project

mgavin committed:
Almost there.
Needs layout cleanup.
Needs keybinding to be better.

mgavin committed:
bakkesmod parsing of strings into its ranges (represented as strings) is bad

mgavin committed:
Debugging rewinding

mgavin committed:
Now with working decoupling from the Steer axis.

mgavin committed:
Line ending and code formatting consistency

mgavin committed:
Removed unused header.

NitrOP7674 commented on CheckpointPlugin.cpp:
```diff
@@ -30,812 +38,1630 @@ int historyTime = 30; // length of history (s)
+*
+*/
+inline void CheckpointPlugin::OpenMenuForKeybinding() {
+    log("HOLY MACKERAL IM TRYING TO CLEAR THESE MOUSE BUTTONS!");
```
Maybe delete this?

mgavin replied:
As for the latest updates and changes, I found that I didn't have to try and clear the mouse buttons being "stuck" anymore (it didn't even work), so I removed all of that stuff.

NitrOP7674 commented on CheckpointPlugin.cpp:
```diff
@@ -30,812 +38,1630 @@ int historyTime = 30; // length of history (s)
+            break;
+        }
+        ++i;
+        // for "runaway" protection... JUST IN CASE THERE'S A BUG!
```
```suggestion
    // Limit 30 for "runaway" protection... JUST IN CASE THERE'S A BUG!
    for (int i = 0; i < 30 && logfile.good(); i++) {
```

NitrOP7674 commented on CheckpointPlugin.cpp:
```diff
@@ -30,812 +38,1630 @@ int historyTime = 30; // length of history (s)
+
+        std::string line = get_prev_line();
+        std::transform(begin(line), end(line), begin(line), [](unsigned char c) { return std::toupper(c); });
+        //log::log_debug("liNE: {}", line);
```
Remove?

mgavin replied:
I didn't want to be constrained by the case of the string that I searched for, 
```
        if (line.contains("CURRENTLY ACTIVE WINDOWS")) {
```

so I put that transforming insurance there... I'll remove the commented out log lines, if that's what you meant.

NitrOP7674 commented on bindings.cpp:
```diff
@@ -52,6 +52,8 @@ void CheckpointPlugin::registerBindingCVars() {
 	cvarManager->registerCvar("cpt_freeze_ball_key", "XboxTypeS_DPad_Up", "Key to bind cpt_freeze_ball to on cpt_apply_bindings");
 	cvarManager->registerCvar("cpt_mirror_state_key", "XboxTypeS_DPad_Down", "Key to bind cpt_mirror_state to on cpt_apply_bindings");
+    cvarManager->registerCvar("cpt_rewind_key", "XboxTypeS_DPad_Down", "Key to bind to cpt_rewind on cpt_apply_bindings");
+    cvarManager->registerCvar("cpt_fastforward_key", "XboxTypeS_DPad_Down", "Key to bind to cpt_fastforward on cpt_apply_bindings");
```
The current defaults are steer L/R - won't this change that?

NitrOP7674 commented on CheckpointPlugin.cpp:
```diff
@@ -30,812 +38,1630 @@ int historyTime = 30; // length of history (s)
+    loadCurCheckpoint();
+}
+
+/**
```
Is it possible to move all the settings UI stuff to another file, maybe Settings.cpp or something?

mgavin replied:
I moved things around to decouple them from the class itself in the latest update. IMGUI stuff stays with the class, but other stuff was just moved outside of it. I moved the IMGUI related stuff (related to the plugin) to another source file. Check whatever commit follows this one (this one 8a68486).

NitrOP7674 commented on CheckpointPlugin.cpp:
```diff
@@ -30,812 +38,1630 @@ int historyTime = 30; // length of history (s)
+        return true; // Staying in rewind; apply state.
+    }
+    rewindState.buttonsDown = buttonsDown;
+    log(std::format("last seek val: {}", rewindState.lastSeekVal));
```
Isn't this going to add a lot of log spam?  Like every 10ms?

mgavin replied:
from my quick testing it was 12ms 😛(240~ fps, idrc to do the math, just reporting the result xD)

I added it to the little canvas stuff that shows up at the top left.

NitrOP7674 commented on CheckpointPlugin.cpp:
```diff
@@ -30,812 +38,1630 @@ int historyTime = 30; // length of history (s)
+void CheckpointPlugin::Render() {
+    if (which_is_being_bound != KEYBIND_ASSIGNWHICH::NONE) {
+        // show the key binding window
+ // creating a translucent background
```
```suggestion
        // create a translucent background
```

NitrOP7674 commented on CheckpointPlugin.h:
```diff
@@ -1,127 +1,220 @@
+            }
+
+            return TRUE;
+            };
```
```suggestion
        };
```

mgavin replied:
removed that section because it was unnecessary

NitrOP7674 commented on CheckpointPlugin.h:
```diff
@@ -1,127 +1,220 @@
+
+        EnumWindows(EnumWindowsFunc, pid);
+        return hWnd;
+        }();
```
```suggestion
    }();
```

mgavin replied:
removed that section because it was unnecessary

NitrOP7674 reviewed (commented):
Thanks for the PR. This sounds really cool! I'll want to try it out for sure - please give me a few days.  In the meantime I have some small comments.  Thanks!

mgavin committed:
Mainly wanted to improve the keybind capturing.
Made refactors of different parts that weren't really related to where they were. So they were put in the pch.
Improved the little popup that appears when binding a key.

mgavin committed:
Made updates and changes with given suggestions in #57

- Removed `cpt_capture_key` because key inputs are captured differently now.
- Set defaults to what they should be given what they were previously
- Put bindings related functions are `bindings.cpp`
- Separated out stuff into a header file `bindings.h`
- Put IMGUI stuff related to the plugin class in its own file `PluginSettings.cpp`

The biggest contentious thing IMO is around `CheckpointPlugin.cpp`L578 where I discriminate against binds that would affect steer and pitch.
